In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
import pandas as pd

df = pd.read_excel('/content/drive/MyDrive/Sample - Superstore Sales (Excel).xlsx')
display(df.head())

,Row ID,Order ID,Order Date,Order Priority,Order Quantity,Sales,Discount,Ship Mode,Profit,Unit Price,...,Zip Code,State,Region,Customer Segment,Product Category,Product Sub-Category,Product Name,Product Container,Product Base Margin,Ship Date
0,1,3,2010-10-13,Low,6,261.5400,0.04,Regular Air,-213.250,38.94,...,60035,Illinois,Central,Small Business,Office Supplies,Storage & Organization,"Eldon Base for stackable storage shelf, platinum",Large Box,0.80,2010-10-20
1,2,6,2012-02-20,Not Specified,2,6.9300,0.01,Regular Air,-4.640,2.08,...,98026,Washington,West,Corporate,Office Supplies,"Scissors, Rulers and Trimmers",Kleencut® Forged Office Shears by Acme United ...,Small Pack,0.55,2012-02-21
2,3,32,2011-07-15,High,26,2808.0800,0.07,Regular Air,1054.820,107.53,...,98387,Washington,West,Corporate,Furniture,Office Furnishings,Tenex Contemporary Contur Chairmats for Low an...,Medium Box,0.65,2011-07-17
3,4,32,2011-07-15,High,24,1761.4000,0.09,Delivery Truck,-1748.560,70.89,...,98387,Washington,West,Corporate,Furniture,Tables,KI Conference Tables,Jumbo Box,0.72,2011-07-16
4,5,32,2011-07-15,High,23,160.2335,0.04,Regular Air,-85.129,7.99,...,98387,Washington,West,Corporate,Technology,Telephones and Communication,Bell Sonecor JB700 Caller ID,Medium Box,0.60,2011-07-17


In [ ]:
import pandas as pd

# 1. Shape of the DataFrame (rows, columns)
print("DataFrame shape:", df.shape)

# 2. Column names
print("\nColumn Names:")
print(df.columns.tolist())

# 3. First 5 rows
print("\nFirst 5 rows of the dataset:")
display(df.head())

# 4. Last 5 rows
print("\nLast 5 rows of the dataset:")
display(df.tail())

# 5. Data types and non-null counts
print("\nData types and Non-Null Counts:")
display(df.info())

# 6. Descriptive statistics for numerical columns
print("\nDescriptive statistics (numerical columns):")
display(df.describe())

# 7. Descriptive statistics for categorical columns
print("\nDescriptive statistics (categorical columns):")
display(df.describe(include='object'))

# 8. Null value counts
print("\nNull value counts per column:")
display(df.isnull().sum())

# 9. Duplicate rows count
print("\nNumber of duplicate rows:")
print(df.duplicated().sum())

# 10. Unique values per column
print("\nUnique value counts per column:")
display(df.nunique())

# 11. Sample value counts for each categorical column
print("\nValue counts for categorical columns:")
for col in df.select_dtypes(include='object').columns:
    print(f"\nColumn: {col}")
    print(df[col].value_counts().head())

# 12. Quick outlier check for numeric columns
print("\nQuick numeric range check (min/max):")
for col in df.select_dtypes(include=['int64', 'float64']).columns:
    print(f"{col}: Min={df[col].min()}, Max={df[col].max()}")

# 13. Example: Check date columns format
# (Uncomment if you know which columns are dates)
# df['Date'] = pd.to_datetime(df['Date'], errors='coerce')
# print("\nDate column range:", df['Date'].min(), "to", df['Date'].max())


DataFrame shape: (8399, 23)

Column Names:
['Row ID', 'Order ID', 'Order Date', 'Order Priority', 'Order Quantity', 'Sales', 'Discount', 'Ship Mode', 'Profit', 'Unit Price', 'Shipping Cost', 'Customer Name', 'City', 'Zip Code', 'State', 'Region', 'Customer Segment', 'Product Category', 'Product Sub-Category', 'Product Name', 'Product Container', 'Product Base Margin', 'Ship Date']

First 5 rows of the dataset:


,Row ID,Order ID,Order Date,Order Priority,Order Quantity,Sales,Discount,Ship Mode,Profit,Unit Price,...,Zip Code,State,Region,Customer Segment,Product Category,Product Sub-Category,Product Name,Product Container,Product Base Margin,Ship Date
0,1,3,2010-10-13,Low,6,261.5400,0.04,Regular Air,-213.250,38.94,...,60035,Illinois,Central,Small Business,Office Supplies,Storage & Organization,"Eldon Base for stackable storage shelf, platinum",Large Box,0.80,2010-10-20
1,2,6,2012-02-20,Not Specified,2,6.9300,0.01,Regular Air,-4.640,2.08,...,98026,Washington,West,Corporate,Office Supplies,"Scissors, Rulers and Trimmers",Kleencut® Forged Office Shears by Acme United ...,Small Pack,0.55,2012-02-21
2,3,32,2011-07-15,High,26,2808.0800,0.07,Regular Air,1054.820,107.53,...,98387,Washington,West,Corporate,Furniture,Office Furnishings,Tenex Contemporary Contur Chairmats for Low an...,Medium Box,0.65,2011-07-17
3,4,32,2011-07-15,High,24,1761.4000,0.09,Delivery Truck,-1748.560,70.89,...,98387,Washington,West,Corporate,Furniture,Tables,KI Conference Tables,Jumbo Box,0.72,2011-07-16
4,5,32,2011-07-15,High,23,160.2335,0.04,Regular Air,-85.129,7.99,...,98387,Washington,West,Corporate,Technology,Telephones and Communication,Bell Sonecor JB700 Caller ID,Medium Box,0.60,2011-07-17



Last 5 rows of the dataset:


,Row ID,Order ID,Order Date,Order Priority,Order Quantity,Sales,Discount,Ship Mode,Profit,Unit Price,...,Zip Code,State,Region,Customer Segment,Product Category,Product Sub-Category,Product Name,Product Container,Product Base Margin,Ship Date
8394,8395,59971,2011-05-31,Critical,26,437.770,0.01,Express Air,-125.80,15.98,...,90266,California,West,Home Office,Technology,Computer Peripherals,"Imation 3.5"" DS/HD IBM Formatted Diskettes, 50...",Small Pack,0.64,2011-06-02
8395,8396,59971,2011-05-31,Critical,45,514.860,0.08,Regular Air,-10.84,12.28,...,90266,California,West,Home Office,Office Supplies,Paper,Xerox 1881,Small Box,0.38,2011-06-02
8396,8397,59971,2011-05-31,Critical,12,356.270,0.10,Express Air,84.53,30.98,...,90266,California,West,Home Office,Office Supplies,Paper,Xerox 19,Small Box,0.40,2011-06-02
8397,8398,59973,2011-11-15,Low,29,11039.750,0.06,Delivery Truck,2148.97,400.97,...,70458,Louisiana,South,Small Business,Technology,Office Machines,Epson FX-980 Dot Matrix Printer,Jumbo Drum,0.59,2011-11-15
8398,8399,59973,2011-11-15,Low,30,4982.944,0.07,Delivery Truck,693.69,217.85,...,70458,Louisiana,South,Small Business,Furniture,Tables,Chromcraft Bull-Nose Wood Round Conference Tab...,Jumbo Box,0.68,2011-11-19



Data types and Non-Null Counts:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8399 entries, 0 to 8398
Data columns (total 23 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   Row ID                8399 non-null   int64         
 1   Order ID              8399 non-null   int64         
 2   Order Date            8399 non-null   datetime64[ns]
 3   Order Priority        8399 non-null   object        
 4   Order Quantity        8399 non-null   int64         
 5   Sales                 8399 non-null   float64       
 6   Discount              8399 non-null   float64       
 7   Ship Mode             8399 non-null   object        
 8   Profit                8399 non-null   float64       
 9   Unit Price            8399 non-null   float64       
 10  Shipping Cost         8399 non-null   float64       
 11  Customer Name         8399 non-null   object        
 12  City                  8399 non-null   objec

None


Descriptive statistics (numerical columns):


,Row ID,Order ID,Order Date,Order Quantity,Sales,Discount,Profit,Unit Price,Shipping Cost,Zip Code,Product Base Margin,Ship Date
count,8399.000000,8399.000000,8399,8399.000000,8399.000000,8399.000000,8399.000000,8399.000000,8399.000000,8399.000000,8336.000000,8399
mean,4200.000000,29965.179783,2010-12-25 00:40:58.578402048,25.571735,1775.878179,0.049671,181.184422,89.346259,12.838557,52839.139421,0.512513,2010-12-27 01:28:48.634361088
min,1.000000,3.000000,2009-01-01 00:00:00,1.000000,2.240000,0.000000,-14140.701600,0.990000,0.490000,1001.000000,0.350000,2009-01-02 00:00:00
25%,2100.500000,15011.500000,2009-12-20 00:00:00,13.000000,143.195000,0.020000,-83.315000,6.480000,3.300000,30337.000000,0.380000,2009-12-22 12:00:00
50%,4200.000000,29857.000000,2010-12-17 00:00:00,26.000000,449.420000,0.050000,-1.500000,20.990000,6.070000,52732.000000,0.520000,2010-12-18 00:00:00
75%,6299.500000,44596.000000,2012-01-01 12:00:00,38.000000,1709.320000,0.080000,162.748000,85.990000,13.990000,77577.000000,0.590000,2012-01-03 12:00:00
max,8399.000000,59973.000000,2012-12-30 00:00:00,50.000000,89061.050000,0.250000,27220.690000,6783.020000,164.730000,99362.000000,0.850000,2012-12-30 00:00:00
std,2424.726789,17260.883447,NaN,14.481071,3585.050525,0.031823,1196.653326,290.354383,17.264052,28509.535635,0.135589,NaN



Descriptive statistics (categorical columns):


,Order Priority,Ship Mode,Customer Name,City,State,Region,Customer Segment,Product Category,Product Sub-Category,Product Name,Product Container
count,8399,8399,8399,8399,8399,8399,8399,8399,8399,8399,8399
unique,5,3,795,1421,48,4,4,3,17,1263,7
top,High,Regular Air,Darren Budd,Columbus,California,Central,Corporate,Office Supplies,Paper,"Global High-Back Leather Tilter, Burgundy",Small Box
freq,1768,6270,41,31,780,2710,3076,4610,1225,24,4347



Null value counts per column:


,0
Row ID,0
Order ID,0
Order Date,0
Order Priority,0
Order Quantity,0
Sales,0
Discount,0
Ship Mode,0
Profit,0
Unit Price,0



Number of duplicate rows:
0

Unique value counts per column:


,0
Row ID,8399
Order ID,5496
Order Date,1418
Order Priority,5
Order Quantity,50
Sales,8153
Discount,16
Ship Mode,3
Profit,7986
Unit Price,751



Value counts for categorical columns:

Column: Order Priority
Order Priority
High             1768
Low              1720
Not Specified    1672
Medium           1631
Critical         1608
Name: count, dtype: int64

Column: Ship Mode
Ship Mode
Regular Air       6270
Delivery Truck    1146
Express Air        983
Name: count, dtype: int64

Column: Customer Name
Customer Name
Darren Budd       41
Ed Braxton        38
Brad Thomas       35
Carlos Soltero    33
Patrick Jones     30
Name: count, dtype: int64

Column: City
City
Columbus       31
Springfield    31
Auburn         26
Roswell        25
Marion         24
Name: count, dtype: int64

Column: State
State
California    780
Texas         577
Illinois      500
Florida       479
Ohio          396
Name: count, dtype: int64

Column: Region
Region
Central    2710
West       1956
East       1895
South      1838
Name: count, dtype: int64

Column: Customer Segment
Customer Segment
Corporate         3076
Home Office       2032
Consumer          16

In [ ]:
display(df[df['Product Base Margin'].isnull()])

,Row ID,Order ID,Order Date,Order Priority,Order Quantity,Sales,Discount,Ship Mode,Profit,Unit Price,...,Zip Code,State,Region,Customer Segment,Product Category,Product Sub-Category,Product Name,Product Container,Product Base Margin,Ship Date
23,24,134,2012-04-30,Not Specified,11,1132.60,0.01,Regular Air,-310.21000,95.99,...,53005,Wisconsin,Central,Home Office,Office Supplies,Storage & Organization,Safco Industrial Wire Shelving,Large Box,NaN,2012-05-02
97,98,613,2011-06-17,High,22,905.08,0.09,Regular Air,127.70000,42.76,...,60087,Illinois,Central,Corporate,Office Supplies,Storage & Organization,"SAFCO Mobile Desk Side File, Wire Frame",Small Box,NaN,2011-06-18
102,103,643,2011-03-24,High,21,2781.82,0.07,Express Air,-695.26000,138.14,...,60031,Illinois,Central,Corporate,Office Supplies,Storage & Organization,"SAFCO Commercial Wire Shelving, Black",Large Box,NaN,2011-03-25
260,261,1824,2009-05-04,Critical,39,10656.26,0.06,Regular Air,3116.54000,276.20,...,97504,Oregon,West,Corporate,Furniture,Chairs & Chairmats,SAFCO Arco Folding Chair,Large Box,NaN,2009-05-05
279,280,1925,2012-03-04,Critical,7,1874.37,0.08,Regular Air,67.84000,276.20,...,67212,Kansas,Central,Corporate,Furniture,Chairs & Chairmats,SAFCO Arco Folding Chair,Large Box,NaN,2012-03-05
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
7628,7629,54592,2011-06-12,Low,28,1971.56,0.05,Delivery Truck,-15.29000,70.98,...,3045,New Hampshire,East,Small Business,Furniture,Bookcases,"Safco Value Mate Steel Bookcase, Baked Enamel ...",Jumbo Box,NaN,2011-06-16
7699,7700,55202,2011-08-24,Low,33,3093.76,0.07,Express Air,-850.71000,95.99,...,3820,New Hampshire,East,Small Business,Office Supplies,Storage & Organization,Safco Industrial Wire Shelving,Large Box,NaN,2011-08-26
7769,7770,55621,2011-03-15,Low,50,6356.68,0.03,Regular Air,1812.86000,128.24,...,27360,North Carolina,South,Home Office,Furniture,Chairs & Chairmats,SAFCO Folding Chair Trolley,Medium Box,NaN,2011-03-17
7825,7826,55938,2012-07-02,Not Specified,18,1769.90,0.05,Express Air,-468.89000,95.99,...,2018,MA,East,Small Business,Office Supplies,Storage & Organization,Safco Industrial Wire Shelving,Large Box,NaN,2012-07-03


In [ ]:
df['Product Base Margin'] = df['Product Base Margin'].fillna(0)
display(df.isnull().sum())

,0
Row ID,0
Order ID,0
Order Date,0
Order Priority,0
Order Quantity,0
Sales,0
Discount,0
Ship Mode,0
Profit,0
Unit Price,0


# Task
Analyze the sales data in "superstore.csv" to calculate and display the following metrics: total sales, sales growth percentage compared to last week, top 3 products by sales, and top 3 employees by sales.

## Calculate total sales

### Subtask:
Calculate the sum of the 'Sales' column in the DataFrame.


**Reasoning**:
Calculate the sum of the 'Sales' column and store it in the `total_sales` variable.



In [ ]:
total_sales = df['Sales'].sum()
print(f"Total Sales: {total_sales}")

Total Sales: 14915600.824000001


## Calculate sales growth percentage (compared to last week)

### Subtask:
Calculate sales growth percentage (compared to last week).


**Reasoning**:
Convert 'Order Date' to datetime, calculate sales for the last two weeks, and compute sales growth percentage.



In [ ]:
# Convert 'Order Date' to datetime objects (already done in previous cell but keeping for completeness)
df['Order Date'] = pd.to_datetime(df['Order Date'])

# Determine the last week and the week before the last week
max_date = df['Order Date'].max()
last_week_start = max_date - pd.Timedelta(days=6)
week_before_last_week_start = max_date - pd.Timedelta(days=13)

# Filter data for the last week and the week before the last week
sales_last_week_df = df[(df['Order Date'] >= last_week_start) & (df['Order Date'] <= max_date)]
sales_week_before_last_week_df = df[(df['Order Date'] >= week_before_last_week_start) & (df['Order Date'] < last_week_start)]

# Calculate the sum of 'Sales' for each week
sales_last_week = sales_last_week_df['Sales'].sum()
sales_week_before_last_week = sales_week_before_last_week_df['Sales'].sum()

# Calculate sales growth percentage, handling division by zero
if sales_week_before_last_week != 0:
    sales_growth_percentage = ((sales_last_week - sales_week_before_last_week) / sales_week_before_last_week) * 100
else:
    sales_growth_percentage = float('inf') if sales_last_week > 0 else 0

print(f"Sales last week: {sales_last_week:.2f}")
print(f"Sales week before last week: {sales_week_before_last_week:.2f}")
print(f"Sales Growth Percentage (compared to last week): {sales_growth_percentage:.2f}%")

Sales last week: 62240.24
Sales week before last week: 38454.01
Sales Growth Percentage (compared to last week): 61.86%


## Identify top 3 products by sales

### Subtask:
Group the DataFrame by 'Product Name', calculate the sum of 'Sales' for each product, sort the products by sales in descending order, and select the top 3.


**Reasoning**:
Group the DataFrame by 'Product Name', calculate the sum of 'Sales' for each product, sort in descending order, and select the top 3 products.



In [ ]:
top_3_products = df.groupby('Product Name')['Sales'].sum().sort_values(ascending=False).head(3)
display(top_3_products)

,Sales
Product Name,
Global Troy™ Executive Leather Low-Back Tilter,275941.52
Polycom ViewStation™ ISDN Videoconferencing Unit,255303.97
Canon PC940 Copier,210910.68


## Identify top 3 employees by sales

### Subtask:
Group the DataFrame by 'Customer Name', calculate the sum of 'Sales' for each customer, sort the customers by sales in descending order, and select the top 3.


**Reasoning**:
Group the DataFrame by 'Customer Name', calculate the sum of 'Sales', sort the results, select the top 3, and display them.



In [ ]:
top_3_employees = df.groupby('Customer Name')['Sales'].sum().sort_values(ascending=False).head(3)
display(top_3_employees)

,Sales
Customer Name,
Emily Phan,117124.4380
Deborah Brumfield,97433.1355
Roy Skaria,92542.1530


## Display metrics

### Subtask:
Print the calculated metrics in a formatted way.


**Reasoning**:
Print the calculated metrics in a formatted way as requested.



In [ ]:
print(f"Total Sales: ${total_sales:,.2f}")
print(f"Sales Growth Percentage (compared to last week): {sales_growth_percentage:,.2f}%")
print("\nTop 3 Products by Sales:")
display(top_3_products)
print("\nTop 3 Employees by Sales:")
display(top_3_employees)

Total Sales: $14,915,600.82
Sales Growth Percentage (compared to last week): 61.86%

Top 3 Products by Sales:


,Sales
Product Name,
Global Troy™ Executive Leather Low-Back Tilter,275941.52
Polycom ViewStation™ ISDN Videoconferencing Unit,255303.97
Canon PC940 Copier,210910.68



Top 3 Employees by Sales:


,Sales
Customer Name,
Emily Phan,117124.4380
Deborah Brumfield,97433.1355
Roy Skaria,92542.1530


## Summary:

### Data Analysis Key Findings

*   The total sales calculated from the dataset is \$14,915,600.82.
*   The sales growth percentage compared to the previous week is 61.86%.
*   The top 3 products by sales are:
    1.  Global Troy™ Executive Leather Low-Back Tilter
    2.  Polycom ViewStation™ ISDN Videoconferencing Unit
    3.  Canon PC940 Copier
*   The top 3 customers by sales are:
    1.  Emily Phan
    2.  Deborah Brumfield
    3.  Roy Skaria

### Insights or Next Steps

*   Investigate the factors contributing to the significant 61.86\% sales growth compared to the previous week.
*   Analyze the sales performance of the top 3 products and customers to understand their impact and identify potential strategies for further growth.


# Task
Generate an HTML email template with placeholders for "Total Sales", "Growth %", and "Top Products". Ensure the template has a clean and readable layout. Provide the final HTML template as a file.

## Generate html structure

### Subtask:
Create the basic HTML structure for an email.


**Reasoning**:
Create a string variable containing the basic HTML structure for an email with a container div and a header.



In [ ]:
html_template = """
<!DOCTYPE html>
<html>
<head>
    <title>Weekly Sales Report</title>
    <style>
        body {{
            font-family: sans-serif;
            margin: 0;
            padding: 20px;
            background-color: #f4f4f4;
        }}
        .container {{
            max-width: 600px;
            margin: 0 auto;
            background-color: #fff;
            padding: 20px;
            border-radius: 8px;
            box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
        }}
        h1 {{
            color: #333;
            text-align: center;
        }}
    </style>
</head>
<body>
    <div class="container">
        <h1>Weekly Sales Report</h1>
        <!-- Placeholders for content will go here -->
    </div>
</body>
</html>
"""
print(html_template)


<!DOCTYPE html>
<html>
<head>
    <title>Weekly Sales Report</title>
    <style>
        body {{
            font-family: sans-serif;
            margin: 0;
            padding: 20px;
            background-color: #f4f4f4;
        }}
        .container {{
            max-width: 600px;
            margin: 0 auto;
            background-color: #fff;
            padding: 20px;
            border-radius: 8px;
            box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
        }}
        h1 {{
            color: #333;
            text-align: center;
        }}
    </style>
</head>
<body>
    <div class="container">
        <h1>Weekly Sales Report</h1>
        <!-- Placeholders for content will go here -->
    </div>
</body>
</html>



## Add placeholders

### Subtask:
Include placeholders for total sales, growth percentage, and top products within the HTML structure.


**Reasoning**:
Modify the HTML template string to include placeholders for the sales metrics.



In [ ]:
html_template = """
<!DOCTYPE html>
<html>
<head>
    <title>Weekly Sales Report</title>
    <style>
        body {{
            font-family: sans-serif;
            margin: 0;
            padding: 20px;
            background-color: #f4f4f4;
        }}
        .container {{
            max-width: 600px;
            margin: 0 auto;
            background-color: #fff;
            padding: 20px;
            border-radius: 8px;
            box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
        }}
        h1 {{
            color: #333;
            text-align: center;
        }}
    </style>
</head>
<body>
    <div class="container">
        <h1>Weekly Sales Report</h1>
        <p><strong>Total Sales:</strong> {{total_sales}}</p>
        <p><strong>Sales Growth Percentage (compared to last week):</strong> {{growth_percentage}}</p>
        <p><strong>Top 3 Products by Sales:</strong></p>
        <ul>
            <li>{{top_product_1}}</li>
            <li>{{top_product_2}}</li>
            <li>{{top_product_3}}</li>
        </ul>
    </div>
</body>
</html>
"""
print(html_template)


<!DOCTYPE html>
<html>
<head>
    <title>Weekly Sales Report</title>
    <style>
        body {{
            font-family: sans-serif;
            margin: 0;
            padding: 20px;
            background-color: #f4f4f4;
        }}
        .container {{
            max-width: 600px;
            margin: 0 auto;
            background-color: #fff;
            padding: 20px;
            border-radius: 8px;
            box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
        }}
        h1 {{
            color: #333;
            text-align: center;
        }}
    </style>
</head>
<body>
    <div class="container">
        <h1>Weekly Sales Report</h1>
        <p><strong>Total Sales:</strong> {{total_sales}}</p>
        <p><strong>Sales Growth Percentage (compared to last week):</strong> {{growth_percentage}}</p>
        <p><strong>Top 3 Products by Sales:</strong></p>
        <ul>
            <li>{{top_product_1}}</li>
            <li>{{top_product_2}}</li>
            <li>{{top_product_3}}</li>


## Style the template

### Subtask:
Add CSS to style the email template for readability and a clean layout.


**Reasoning**:
Add CSS styles to the HTML template for better readability and a clean layout, focusing on paragraphs and list items.



In [ ]:
html_template = """
<!DOCTYPE html>
<html>
<head>
    <title>Weekly Sales Report</title>
    <style>
        body {{
            font-family: sans-serif;
            margin: 0;
            padding: 20px;
            background-color: #f4f4f4;
            color: #333;
        }}
        .container {{
            max-width: 600px;
            margin: 20px auto;
            background-color: #fff;
            padding: 20px;
            border-radius: 8px;
            box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
        }}
        h1 {{
            color: #333;
            text-align: center;
            margin-top: 0;
        }}
        p {{
            font-size: 16px;
            line-height: 1.5;
            margin-bottom: 10px;
        }}
        strong {{
            color: #555;
        }}
        ul {{
            list-style-type: disc;
            margin-top: 10px;
            padding-left: 20px;
        }}
        li {{
            margin-bottom: 5px;
        }}
    </style>
</head>
<body>
    <div class="container">
        <h1>Weekly Sales Report</h1>
        <p><strong>Total Sales:</strong> {total_sales}</p>
        <p><strong>Sales Growth Percentage (compared to last week):</strong> {growth_percentage}</p>
        <p><strong>Top 3 Products by Sales:</strong></p>
        <ul>
            <li>{top_product_1}</li>
            <li>{top_product_2}</li>
            <li>{top_product_3}</li>
        </ul>
    </div>
</body>
</html>
"""
print(html_template)


<!DOCTYPE html>
<html>
<head>
    <title>Weekly Sales Report</title>
    <style>
        body {{
            font-family: sans-serif;
            margin: 0;
            padding: 20px;
            background-color: #f4f4f4;
            color: #333;
        }}
        .container {{
            max-width: 600px;
            margin: 20px auto;
            background-color: #fff;
            padding: 20px;
            border-radius: 8px;
            box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
        }}
        h1 {{
            color: #333;
            text-align: center;
            margin-top: 0;
        }}
        p {{
            font-size: 16px;
            line-height: 1.5;
            margin-bottom: 10px;
        }}
        strong {{
            color: #555;
        }}
        ul {{
            list-style-type: disc;
            margin-top: 10px;
            padding-left: 20px;
        }}
        li {{
            margin-bottom: 5px;
        }}
    </style>
</head>
<body>
    <div cla

## Populate with data

### Subtask:
Generate Python code to populate the HTML template with the calculated metrics from the previous analysis.


**Reasoning**:
Define the variables with the calculated metrics and format them as strings, then use the format method to populate the html template.



In [ ]:
total_sales_formatted = f"${total_sales:,.2f}"
sales_growth_percentage_formatted = f"{sales_growth_percentage:,.2f}%"
top_products_list = top_3_products.index.tolist()

populated_html = html_template.format(
    total_sales=total_sales_formatted,
    growth_percentage=sales_growth_percentage_formatted,
    top_product_1=top_products_list[0],
    top_product_2=top_products_list[1],
    top_product_3=top_products_list[2]
)

print(populated_html)


<!DOCTYPE html>
<html>
<head>
    <title>Weekly Sales Report</title>
    <style>
        body {
            font-family: sans-serif;
            margin: 0;
            padding: 20px;
            background-color: #f4f4f4;
            color: #333;
        }
        .container {
            max-width: 600px;
            margin: 20px auto;
            background-color: #fff;
            padding: 20px;
            border-radius: 8px;
            box-shadow: 0 0 10px rgba(0, 0, 0, 0.1);
        }
        h1 {
            color: #333;
            text-align: center;
            margin-top: 0;
        }
        p {
            font-size: 16px;
            line-height: 1.5;
            margin-bottom: 10px;
        }
        strong {
            color: #555;
        }
        ul {
            list-style-type: disc;
            margin-top: 10px;
            padding-left: 20px;
        }
        li {
            margin-bottom: 5px;
        }
    </style>
</head>
<body>
    <div class="container"

**Reasoning**:
The previous code failed because the variables `total_sales`, `sales_growth_percentage`, and `top_3_products` were not defined in the current session. I need to re-execute the cells that define these variables before attempting to populate the HTML template.



In [ ]:
with open('weekly_sales_report.html', 'w') as f:
    f.write(populated_html)

print("HTML template saved as 'weekly_sales_report.html'")

HTML template saved as 'weekly_sales_report.html'


## Summary:

The HTML email template has been generated and saved as `weekly_sales_report.html`. You can download this file and use it as a template for sending out weekly sales reports with the calculated metrics.

# Task
Automate Email Sending - Configure SMTP in Python to send emails. - Write a Python function that sends the email with the summary report. - Test sending emails to their personal/test email accounts. - Deliverable: A Python script that sends a sample email with dummy metrics.

## Configure smtp settings

### Subtask:
Set up the necessary SMTP server details and login credentials.


**Reasoning**:
Define the variables for SMTP server details and login credentials.



In [ ]:
smtp_server = 'smtp.gmail.com'  # e.g., 'smtp.gmail.com'
smtp_port = 587  # e.g., 587 for TLS
sender_email = 'your_email'
sender_password = 'your password' # Be cautious with storing passwords directly, consider using environment variables or a secure method

## Create email content

### Subtask:
Prepare the email subject, body (using the HTML template), and sender/recipient addresses.


**Reasoning**:
Define the email recipient, subject, and use the previously generated populated HTML as the email body.



In [ ]:
recipient_email = 'walimokhan@gmail.com' # Replace with your recipient email
email_subject = 'Weekly Sales Report'
email_body = populated_html

## Write a function to send email

### Subtask:
Create a Python function that takes the email details and sends the email using the configured SMTP settings.


**Reasoning**:
Import necessary libraries and define the function to send the email.



In [ ]:
import smtplib
from email.mime.text import MIMEText

def send_sales_report_email(recipient_email, subject, html_body):
    """Sends a weekly sales report email with HTML content.

    Args:
        recipient_email: The email address of the recipient.
        subject: The subject of the email.
        html_body: The HTML content of the email body.
    """
    msg = MIMEText(html_body, 'html')
    msg['Subject'] = subject
    msg['From'] = sender_email
    msg['To'] = recipient_email

    try:
        with smtplib.SMTP(smtp_server, smtp_port) as server:
            server.starttls()  # Secure the connection
            server.login(sender_email, sender_password)
            server.sendmail(sender_email, recipient_email, msg.as_string())
        print("Email sent successfully!")
    except Exception as e:
        print(f"Failed to send email: {e}")


## Test sending email

### Subtask:
Test the function by sending a sample email to a personal/test email account.


**Reasoning**:
Call the send_sales_report_email function with the defined variables and placeholder SMTP credentials to test sending an email.



In [ ]:
# Replace with your actual SMTP server details, sender email address, and sender password
smtp_server = 'smtp.gmail.com'
smtp_port = 587  # e.g., 587 for TLS
sender_email = 'your email'
sender_password = 'your password' # Be cautious with storing passwords directly, consider using environment variables or a secure method

# Replace with your recipient email
recipient_email = 'walimokhan@gmail.com'
email_subject = 'Weekly Sales Report' # Changed subject back to the original
# Use the populated_html variable which contains the real data
email_body = populated_html

send_sales_report_email(recipient_email, email_subject, email_body)

Email sent successfully!


## Summary:

### Data Analysis Key Findings

*   The process successfully defined variables for SMTP server details, sender email, sender password, recipient email, and email subject.
*   A Python function `send_sales_report_email` was created using `smtplib` and `email.mime.text.MIMEText` to send emails with HTML content, including TLS encryption and authentication.
*   Attempting to send a test email failed with an `[Errno -2] Name or service not known` error because the placeholder SMTP server details were not replaced with actual, valid information.

### Insights or Next Steps

*   Replace the placeholder values for `smtp_server`, `smtp_port`, `sender_email`, and `sender_password` with actual SMTP configuration details to successfully send test emails.
*   Implement a more secure method for storing and accessing sensitive credentials like the sender password, such as using environment variables or a configuration file, rather than hardcoding them directly in the script.
